File: src/game/entities/Actor.ts
import { Container, Sprite } from 'pixi.js';
import { Animator } from '../animation/Animator';
import { PLAYFIELD } from '../config';
import type { AnimationBank, Rect, Vec2 } from '../types';
import { clamp } from '../../utils/math';
import { integrateHorizontalLaunch, integrateLaunch } from '../combat/knockdownPhysics';
import { clampFeetX, horizontalExtents } from '../animation/visualBounds';

let nextActorId = 1;

export interface HitResult {
  accepted: boolean;
  killed: boolean;
  knockedDown: boolean;
}

export class Actor {
  readonly actorId = nextActorId++;
  readonly root = new Container();
  readonly sprite = new Sprite();
  readonly transitionSprite = new Sprite();
  readonly outlineSprites: Sprite[] = [];
  readonly animator: Animator;
  readonly position: Vec2;
  velocity: Vec2 = { x: 0, y: 0 };
  maxHealth: number;
  health: number;
  readonly visualScale: number;
  facing: -1 | 1 = 1;
  state = 'idle';
  stateElapsed = 0;
  invulnerable = 0;
  hitFlash = 0;
  alpha255 = 255;
  elevation = 0;
  verticalVelocity = 0;
  landedThisFrame = false;
  landingImpact = 0;
  private remainingKnockdownBounces = 0;
  private playfieldBounds: { left: number; right: number; top: number; bottom: number } = { ...PLAYFIELD };
  private playfieldProfile?: (worldX: number) => [number, number];
  dead = false;
  removeReady = false;

  constructor(bank: AnimationBank, position: Vec2, maxHealth: number, visualScale = 1) {
    this.animator = new Animator(bank);
    this.position = { ...position };
    this.maxHealth = maxHealth;
    this.health = maxHealth;
    this.visualScale = clamp(visualScale, 0.85, 1.15);

    this.root.sortableChildren = true;
    for (const [dx, dy] of [[-1, 0], [1, 0], [0, -1], [0, 1]] as const) {
      const outline = new Sprite(this.animator.frame.texture);
      outline.anchor.set(0.5, 1);
      outline.tint = 0x181212;
      outline.alpha = 0.58;
      outline.position.set(dx, dy + this.animator.frame.offsetY);
      outline.zIndex = 0;
      this.outlineSprites.push(outline);
      this.root.addChild(outline);
    }
    this.transitionSprite.anchor.set(0.5, 1);
    this.transitionSprite.zIndex = 0.5;
    this.transitionSprite.visible = false;
    this.root.addChild(this.transitionSprite);
    this.sprite.texture = this.animator.frame.texture;
    this.sprite.anchor.set(0.5, 1);
    this.sprite.zIndex = 1;
    this.root.addChild(this.sprite);
    this.syncVisual(true);
  }

  get feet(): Vec2 {
    return this.position;
  }

  get collisionRadius(): Vec2 {
    if (this.state === 'knockdown' || this.state === 'dead') return { x: 60, y: 16 };
    if (this.state === 'getup') return { x: 38, y: 19 };
    return { x: 35, y: 21 };
  }

  get hurtbox(): Rect {
    let width = 58;
    let height = 142;
    if (this.state === 'knockdown' || this.state === 'dead') {
      width = 126;
      height = 46;
    } else if (this.state === 'getup') {
      width = 76;
      height = 102;
    }
    return {
      x: this.position.x - width / 2,
      y: this.position.y - this.elevation - height,
      width,
      height,
    };
  }

  get canBeHit(): boolean {
    return !this.dead && this.invulnerable <= 0 && !['knockdown', 'getup'].includes(this.state);
  }

  beginState(state: string, animation = state, restart = true): void {
    this.state = state;
    this.stateElapsed = 0;
    this.animator.setPlaybackRate(1);
    this.animator.play(animation, restart);
    this.syncVisual(true);
  }

  updateCommon(dt: number): void {
    this.landedThisFrame = false;
    this.landingImpact = 0;
    this.stateElapsed += dt;
    this.invulnerable = Math.max(0, this.invulnerable - dt);
    this.hitFlash = Math.max(0, this.hitFlash - dt);
    const changed = this.animator.update(dt);
    const fallen = ['knockdown', 'dead'].includes(this.state);
    const launched = fallen && (this.verticalVelocity !== 0 || this.elevation > 0);
    if (fallen) {
      const horizontal = integrateHorizontalLaunch(this.position.x, this.velocity.x, dt, launched);
      this.position.x = horizontal.position;
      this.velocity.x = horizontal.velocity;
      this.position.y += this.velocity.y * dt;
      this.velocity.y *= Math.max(0, 1 - 3.8 * dt);
    } else {
      this.position.x += this.velocity.x * dt;
      this.position.y += this.velocity.y * dt;
      const damping = Math.max(0, 1 - 8 * dt);
      this.velocity.x *= damping;
      this.velocity.y *= damping;
    }
    if (['knockdown', 'dead'].includes(this.state) && (this.verticalVelocity !== 0 || this.elevation > 0)) {
      const impactVelocity = Math.abs(this.verticalVelocity);
      const launch = integrateLaunch(this.elevation, this.verticalVelocity, dt);
      this.elevation = launch.elevation;
      this.verticalVelocity = launch.verticalVelocity;
      this.landedThisFrame = launch.landed;
      if (launch.landed) {
        this.landingImpact = impactVelocity;
        if (this.remainingKnockdownBounces > 0 && impactVelocity >= 285) {
          this.remainingKnockdownBounces -= 1;
          this.elevation = 1;
          this.verticalVelocity = impactVelocity * 0.31;
          this.velocity.x *= 0.74;
          this.velocity.y *= 0.58;
        } else {
          this.velocity.x *= 0.42;
          this.velocity.y *= 0.42;
        }
      }
    }
    this.syncVisual(changed);
  }

  clampToPlayfield(): void {
    const flip = this.facing === this.animator.sourceFacing ? 1 : -1;
    this.position.x = clampFeetX(
      this.position.x,
      this.animator.frame,
      flip,
      this.playfieldBounds.left,
      this.playfieldBounds.right,
      6,
      this.visualScale,
    );
    const profile = this.playfieldProfile?.(this.position.x);
    const top = profile ? Math.min(profile[0], profile[1]) : this.playfieldBounds.top;
    const bottom = profile ? Math.max(profile[0], profile[1]) : this.playfieldBounds.bottom;
    this.position.y = clamp(this.position.y, top, bottom);
  }

  setPlayfieldBounds(left: number, right: number, top: number = PLAYFIELD.top, bottom: number = PLAYFIELD.bottom): void {
    this.playfieldBounds = { left, right, top, bottom };
    this.clampToPlayfield();
  }

  setPlayfieldProfile(profile?: (worldX: number) => [number, number]): void {
    this.playfieldProfile = profile;
    this.clampToPlayfield();
  }

  visualHorizontalBounds(): { left: number; right: number } {
    const flip = this.facing === this.animator.sourceFacing ? 1 : -1;
    const extents = horizontalExtents(this.animator.frame, flip, this.visualScale);
    return { left: this.position.x + extents.left, right: this.position.x + extents.right };
  }

  receiveHit(damage: number, knockback: Vec2, knockdown = false, launchVelocity = 0): HitResult {
    if (!this.canBeHit) return { accepted: false, killed: false, knockedDown: false };
    this.health = Math.max(0, this.health - damage);
    this.hitFlash = 0.075;
    this.invulnerable = 0.12;
    this.velocity = { ...knockback };
    const killed = this.health <= 0;
    if (killed) {
      this.remainingKnockdownBounces = launchVelocity >= 560 ? 2 : launchVelocity >= 420 ? 1 : 0;
      this.elevation = launchVelocity > 0 ? 1 : 0;
      this.verticalVelocity = Math.max(0, launchVelocity);
      this.dead = true;
      this.beginState('dead', 'dead');
      this.invulnerable = 999;
      return { accepted: true, killed: true, knockedDown: true };
    }
    if (knockdown) {
      this.remainingKnockdownBounces = launchVelocity >= 560 ? 2 : launchVelocity >= 420 ? 1 : 0;
      this.elevation = launchVelocity > 0 ? 1 : 0;
      this.verticalVelocity = Math.max(0, launchVelocity);
      this.beginState('knockdown', 'knockdown');
      this.invulnerable = 0.55;
      return { accepted: true, killed: false, knockedDown: true };
    }
    this.remainingKnockdownBounces = 0;
    this.beginState('hit', 'hit');
    return { accepted: true, killed: false, knockedDown: false };
  }

  visualTop(): number {
    const frame = this.animator.frame;
    const [, top, , height] = frame.bounds;
    const scale = frame.scale * this.visualScale;
    const canvasBottom = this.position.y - this.elevation + frame.offsetY * this.visualScale;
    const canvasTop = canvasBottom - frame.height * scale;
    return height > 0 ? canvasTop + top * scale : canvasTop;
  }

  syncVisual(forceTexture = false): void {
    const frame = this.animator.frame;
    if (forceTexture || this.sprite.texture !== frame.texture) {
      this.sprite.texture = frame.texture;
      for (const outline of this.outlineSprites) outline.texture = frame.texture;
    }
    const flip = this.facing === this.animator.sourceFacing ? 1 : -1;
    const visualFrameScale = frame.scale * this.visualScale;
    this.sprite.scale.set(flip * visualFrameScale, visualFrameScale);
    this.sprite.x = frame.offsetX * visualFrameScale;
    this.sprite.y = frame.offsetY * this.visualScale;
    this.sprite.tint = this.hitFlash > 0 ? 0xffa08e : 0xffffff;
    const transition = this.animator.transitionFrame;
    if (transition) {
      const transitionFlip = this.facing === this.animator.transitionSourceFacing ? 1 : -1;
      this.transitionSprite.texture = transition.texture;
      const transitionScale = transition.scale * this.visualScale;
      this.transitionSprite.scale.set(transitionFlip * transitionScale, transitionScale);
      this.transitionSprite.position.set(transition.offsetX * transitionScale, transition.offsetY * this.visualScale);
      this.transitionSprite.alpha = this.animator.transitionAlpha;
      this.transitionSprite.tint = this.sprite.tint;
      this.transitionSprite.visible = true;
      this.sprite.alpha = 1 - this.animator.transitionAlpha;
    } else {
      this.transitionSprite.visible = false;
      this.sprite.alpha = 1;
    }
    for (let i = 0; i < this.outlineSprites.length; i += 1) {
      const outline = this.outlineSprites[i]!;
      const offsets: ReadonlyArray<readonly [number, number]> = [[-1, 0], [1, 0], [0, -1], [0, 1]];
      const [dx, dy] = offsets[i] ?? [0, 0];
      outline.scale.set(flip * visualFrameScale, visualFrameScale);
      outline.x = frame.offsetX * visualFrameScale + dx;
      outline.y = frame.offsetY * this.visualScale + dy;
      outline.alpha = 0.58 * this.sprite.alpha;
    }
    this.root.alpha = this.alpha255 / 255;
    this.root.position.set(this.position.x, this.position.y - this.elevation);
    this.root.zIndex = Math.round(this.position.y);
  }

  destroy(): void {
    this.root.destroy({ children: true });
  }
}

File: legacy-reference/python/palermo_streets/entities/actor.py
from __future__ import annotations

from dataclasses import dataclass
from itertools import count
import pygame

from ..systems.animation import AnimationBank, Animator


_IDS = count(1)


@dataclass(frozen=True, slots=True)
class AttackData:
    name: str
    damage: int
    startup: float
    active: float
    recovery: float
    range_x: int
    range_y: int
    knockback_x: float
    knockback_y: float
    fury_gain: int
    hit_stop: float
    shake: float
    knockdown: bool = False
    multi_hit: bool = False

    @property
    def total(self) -> float:
        return self.startup + self.active + self.recovery


@dataclass(frozen=True, slots=True)
class HitResult:
    accepted: bool
    killed: bool
    knocked_down: bool


class Actor:
    def __init__(
        self,
        animations: AnimationBank,
        position: pygame.Vector2,
        max_health: int,
        *,
        visual_height: int,
    ) -> None:
        self.actor_id = next(_IDS)
        self.animator = Animator(animations)
        self.position = position
        self.velocity = pygame.Vector2()
        self.max_health = max_health
        self.health = max_health
        self.visual_height = visual_height
        # facing è la direzione desiderata nel mondo. L'orientamento nativo
        # viene letto dalla singola AnimationClip, perché il Marco Pack
        # contiene clip sorgente rivolte sia a sinistra sia a destra.
        self.facing = 1
        self.state = "idle"
        self.state_elapsed = 0.0
        self.invulnerable = 0.0
        self.hit_flash = 0.0
        self.alpha = 255
        self.dead = False
        self.remove_ready = False

    @property
    def feet(self) -> pygame.Vector2:
        return self.position

    @property
    def frame_image(self) -> pygame.Surface:
        return self.animator.frame.image

    @property
    def collision_radius(self) -> pygame.Vector2:
        """Ingombro sul piano stradale, indipendente dal canvas trasparente."""
        if self.state in ("knockdown", "dead"):
            return pygame.Vector2(60.0, 16.0)
        if self.state == "getup":
            return pygame.Vector2(38.0, 19.0)
        return pygame.Vector2(35.0, 21.0)

    @property
    def hurtbox(self) -> pygame.Rect:
        # La hitbox fisica non dipende dal canvas trasparente dello sprite.
        if self.state in ("knockdown", "dead"):
            width, height = 126, 46
        elif self.state == "getup":
            width, height = 76, 102
        else:
            width, height = 58, 142
        return pygame.Rect(
            int(self.position.x - width / 2),
            int(self.position.y - height),
            width,
            height,
        )

    @property
    def can_be_hit(self) -> bool:
        return not self.dead and self.invulnerable <= 0.0 and self.state not in ("knockdown", "getup")

    def begin_state(self, state: str, animation: str | None = None, *, restart: bool = True) -> None:
        self.state = state
        self.state_elapsed = 0.0
        self.animator.play(animation or state, restart=restart)

    def update_common(self, dt: float) -> None:
        self.state_elapsed += dt
        self.invulnerable = max(0.0, self.invulnerable - dt)
        self.hit_flash = max(0.0, self.hit_flash - dt)
        self.animator.update(dt)
        self.position += self.velocity * dt
        self.velocity *= max(0.0, 1.0 - 8.0 * dt)

    def clamp_to_playfield(self, playfield: tuple[float, float, float, float]) -> None:
        left, top, right, bottom = playfield
        self.position.x = max(left, min(right, self.position.x))
        self.position.y = max(top, min(bottom, self.position.y))

    def receive_hit(
        self,
        damage: int,
        knockback: pygame.Vector2,
        *,
        knockdown: bool = False,
    ) -> HitResult:
        if not self.can_be_hit:
            return HitResult(False, False, False)

        self.health = max(0, self.health - damage)
        self.hit_flash = 0.075
        self.invulnerable = 0.12
        self.velocity = knockback
        killed = self.health <= 0
        if killed:
            self.dead = True
            self.begin_state("dead", "dead")
            self.invulnerable = 999.0
            return HitResult(True, True, True)

        if knockdown:
            self.begin_state("knockdown", "knockdown")
            self.invulnerable = 0.55
            return HitResult(True, False, True)

        self.begin_state("hit", "hit")
        return HitResult(True, False, False)

    def render_image(self) -> pygame.Surface:
        """Restituisce il frame orientato, preservando l'alpha trasparente."""
        image = self.animator.frame.image
        if self.facing != self.animator.source_facing:
            image = pygame.transform.flip(image, True, False)
        if self.hit_flash > 0.0:
            flashed = image.copy()
            # Alpha 0: colora solo i pixel esistenti e non crea il rettangolo
            # rosso del canvas trasparente osservato nel video v0.5.3.
            flashed.fill((255, 95, 80, 0), special_flags=pygame.BLEND_RGBA_ADD)
            image = flashed
        if self.alpha < 255:
            image = image.copy()
            image.set_alpha(self.alpha)
        return image

    def visual_top(self, offset: pygame.Vector2 = pygame.Vector2()) -> int:
        """Y superiore della silhouette visibile nel mondo/schermo.

        Il canvas trasparente è molto più alto del personaggio: usare questo
        valore evita barre vita troppo alte nelle pose accovacciate o a terra.
        """
        frame = self.animator.frame
        _left, top, _width, height = frame.bounds
        if height <= 0:
            bbox = frame.image.get_bounding_rect(min_alpha=8)
            top = bbox.top
        canvas_bottom = round(self.position.y + offset.y + frame.offset_y)
        canvas_top = canvas_bottom - frame.image.get_height()
        return canvas_top + top

    def draw(self, target: pygame.Surface, offset: pygame.Vector2 = pygame.Vector2()) -> None:
        frame = self.animator.frame
        image = self.render_image()
        x = round(self.position.x + offset.x + frame.offset_x * self.facing)
        y = round(self.position.y + offset.y + frame.offset_y)
        rect = image.get_rect(midbottom=(x, y))

        # Sottile outline scuro stile arcade. È precalcolato una sola volta
        # durante il caricamento della AnimationBank, non a ogni draw.
        outline = frame.outline
        if outline is not None:
            if self.facing != self.animator.source_facing:
                outline = pygame.transform.flip(outline, True, False)
            if self.alpha < 255:
                outline = outline.copy()
                outline.set_alpha(self.alpha)
            for dx, dy in ((-1, 0), (1, 0), (0, -1), (0, 1)):
                target.blit(outline, rect.move(dx, dy))
        target.blit(image, rect)
